Add a french article introducing k8s crossplane (#1112)
eleven-labs/blog.eleven-labs.com: _articles/fr/2024-04-30-k8s-crossplane-introduction.md

---
contentType: article
lang: fr
date: '2024-04-30'
slug: k8s-crossplane-introduction
title: 'Gestion de ressources externes et platform engineering depuis un cluster Kubernetes avec Crossplane'
excerpt: >-

>   Découvrez Crossplane 🍦, un outil open-source qui étends Kubernetes avec des concepts de platform engineering et permet de gérer le cycle de vie de vos ressources cloud dans une boucle de réconciliation (contrairement à Terraform).

Reviewer: Il faudrait ajouter les balises SEO : 
seo:
    title: Utiliser Crossplane dans un cluster Kubernetes : mode d'emploi
    description: Découvrez ce qu'est crossplane et apprenez grâce à notre mode d'emploi à le configurer et à l'utiliser étapes par étapes dans votre projet.

cover:
  path: /imgs/articles/2024-04-30-k8s-crossplane-introduction/cover.jpg
categories:
  - architecture
authors:
  - dfert
keywords:
  - kubernetes
  - k8s
  - cloud

>   - platform enineering

Reviewer: ```suggestion
  - platform enineering
  - crossplane
```

```

On pourra également installer la CLI `crank` disponible à cette adresse : https://releases.crossplane.io/stable/current/bin. Elle contient des outils pour le rendering, la validation de templates et des fonctions en local.

Une fois que tout est en place, rentrons maintenant dans le vif du sujet.

## En profondeur

### Providers

A l'image de Terraform, Crossplane utilise la notion de _provider_ pour désigner le composant responsable de faire les appels API vers les systèmes externes. Les paquets fournits par UpBound sont distribués l'échelle d'un service cloud (i.e. un provider pour S3, un autre pour RDS, etc.) et à l'échelle de tous les services également (i.e. un seul controller capable de gérer tous les services d'un cloud provider). UpBound distribue ces providers sur son [marketplace](https://marketplace.upbound.io/providers) sur lequel on pourra consulter la liste de tous les providers natifs.

Exemple de déclaration d'un provider AWS :

```yaml
apiVersion: pkg.crossplane.io/v1
kind: Provider
metadata:
  name: provider-family-aws
spec:
  package: xpkg.upbound.io/upbound/provider-family-aws:v1.3.1
```

> L'ajout d'une provider créé de nouveaux enpoints d'API qu'on utilisera pour créer nos différentes ressources externes. Les endpoints sont propres à chaque provider, et on consultera la documentation du provider pour voir quels type d'objets nous sont désormais accessibles. Le provider RDS rajoute l'endpoint `rds.aws.upbound.io/v1beta1` par exemple. Notez que Crossplane peut gérer des révisions lorsque l'on change la version du paquet d'un provider, mais par défault, il utilisera toujours la plus récente et n'aura qu'une profondeur d'historique de 1.

Reviewer: ```suggestion
L'ajout d'une provider créé de nouveaux enpoints d'API qu'on utilisera pour créer nos différentes ressources externes. Les endpoints sont propres à chaque provider, et on consultera la documentation du provider pour voir quels type d'objets nous sont désormais accessibles. Le provider RDS rajoute l'endpoint `rds.aws.upbound.io/v1beta1` par exemple. Notez que Crossplane peut gérer des révisions lorsque l'on change la version du paquet d'un provider, mais par défaut, il utilisera toujours la plus récente et n'aura qu'une profondeur d'historique de 1.
```

On va devoir configurer ce provider maintenant (pour les crédentials notamment). Cela peut se faire à deux niveaux :

> - Niveau global — En créant un objet `DeploymentRuntimeConfig` qu'on référencera dans le champ `spec.runtimeConfigRef` de l'objet `Provider`. La configuration sera alors appliquée au runtime du paquet du provider

Reviewer: ```suggestion
- Niveau global — En créant un objet `DeploymentRuntimeConfig` qu'on référencera dans le champ `spec.runtimeConfigRef` de l'objet `Provider`. La configuration sera alors appliquée au runtime du paquet du provider.
```

```yaml
apiVersion: pkg.crossplane.io/v1beta1
kind: DeploymentRuntimeConfig
metadata:
  name: enable-ess
spec:
  deploymentTemplate:
    spec:
    # etc.
---
apiVersion: pkg.crossplane.io/v1
kind: Provider
metadata:
  name: provider-family-aws
spec:
  package: xpkg.upbound.io/upbound/provider-family-aws:v1.3.1
  runtimeConfigRef:
    name: enable-ess
    # etc.
```

- Niveau *ressource managée* — En créant un objet `ProviderConfig` qu'on référencera dans les ressources managées via `spec.providerConfigRe.name` (exemple dans la prochaine section.) Ce sera alors une configuration que le controlleur n'appliquera que pour la ressource en question. Comme les ressources créés depuis l'endpoints RDS sont *defacto* gérées par le Provider RDS, on ne précisera pas ici de reférence entre les objets `ProviderConfig` et `Provider`.

```yaml
apiVersion: aws.upbound.io/v1beta1
kind: ProviderConfig
metadata:
  name: provider-family-aws
spec:
  credentials:
    secretRef:
      key: credentials
      name: example-aws-creds
      namespace: crossplane-system
    source: Secret
```

> ### Créer des ressources externes

Reviewer: ```suggestion
### Créer des ressources externes avec Crossplane
```

Maintenant que notre provider RDS est déclaré et configuré, Crossplane va pouvoir interragir avec l'API AWS pour créer nos ressources. Demandons par exemple une instance RDS. On créé un objet Kubernetes qu'on appellera alors une _ressource managée_ :

```yaml
apiVersion: rds.aws.upbound.io/v1beta1
kind: Instance
metadata:
  name: my-rds-instance
spec:
  deletionPolicy: Delete
  providerConfigRef: 
    name: provider-family-aws
  forProvider:
    engine: postgres
    engineVersion: 16.1
    storageType: gp3
    # etc.
```
N.B : La documentation de cette API se trouve ici : https://marketplace.upbound.io/providers/upbound/provider-aws-rds/v1.3.1)

> Une fois que cette ressource managée est déclarée, Crossplane va alors scrutter le cloud provider pour vérifier que la ressource externe est dans l'état que nous avons demandé et tentera de corriger le *drift* sinon. Par défaut, Crossplane interroge les providers toutes les minutes et on pourra ajuster cet interval avec l'argument `--poll-interval n` du controller. Notez qu'avec une valeur trop faible nos appels API risquent d'être bloqués par les clouds providers car trop fréquents. Crossplane reportera le status des ressources managées en injectant un bloc `Conditions` directement dans l'objet. Aussi, toutes les informations de la ressource *externe* seront dans `status.atProvider` (e.g. ARN, id, etc.)

Reviewer: ```suggestion
Une fois que cette ressource managée est déclarée, Crossplane va alors scruter le cloud provider pour vérifier que la ressource externe est dans l'état demandé et tentera de corriger le *drift* sinon. Par défaut, Crossplane interroge les providers toutes les minutes mais il est possible d'ajuster cet intervalle avec l'argument `--poll-interval n` du controller. Notez qu'avec une valeur trop faible nos appels API risquent d'être bloqués par les clouds providers car trop fréquents. Crossplane reportera le statut des ressources managées en injectant un bloc `Conditions` directement dans l'objet. Aussi, toutes les informations de la ressource *externe* seront dans `status.atProvider` (e.g. ARN, id, etc.)
```

Sur le cloud, certaines ressources externes ont des attributs immutables. Par exemple, AWS ne permet pas de modifier le nom d'un bucket S3. Terraform triche un peu avec ça en autorisant la modification des champs immutables mais en proposant alors de recréer ces ressources (i.e. supprimer le bucket et le recréer avec le nouveau nom). Crossplane, ne dispose pas de ce type de mécanisme et on ne pourra pas modifier les champs immutables. Pour faire l'équivalent, on devra supprimer et recréer nous même l'objet en question. Le nom de la ressource externe sera le même que celui de la ressource managée par défaut. 

> ### Policies

Reviewer: ```suggestion
### Les policies de Crossplane
```

![police no way]({BASE_URL}/imgs/articles/2024-04-30-k8s-crossplane-introduction/policies_resized.png)

#### Deletion Policies

Crossplane permet de configurer comment le provider se comporte quand une ressource managée est supprimée de Kubernetes. Cette configuration est faite dans le bloc `spec.deletionPolicy` de la ressource et on distingue deux cas :

- `deletionPolicy: Delete` — La ressource externe est supprimée quand on l'objet est supprimé de Kubernetes (comportement par défaut)
- `deletionPolicy: Orphan` — La ressource externe n'est pas supprimée quand l'objet est supprimé de Kubernetes (ce qui laisse des orphelins sur le cloud, à l'image d'un state RM pour Terraform)

#### Management Policies

> Les `managementPolicies` permettent de limiter les droits de Crossplane sur les ressources externes. On les spécifies dans le tableau `spec.managementPolicies` de nos managed ressources. Voici les différentes valeurs possibles (elles sont cumulables)

Reviewer: ```suggestion
Les `managementPolicies` permettent de limiter les droits de Crossplane sur les ressources externes. On les spécifie dans le tableau `spec.managementPolicies` de nos managed ressources. Voici les différentes valeurs possibles (elles sont cumulables)
```

</br>

| Policy          | Fonctionnement                                                                                                                                                              |
|---------------- |---------------------------------------------------------------------------------------------------------------------------------------------------------------------------  |
| *               | (Defaut) Crossplane a tous les droits sur la ressource externe.                                                                                                                |
| Create          | Crossplane est autorisé à créer la ressource externe si elle n'existe pas.                                                                       |
| Delete          | Crossplane est autorisé à supprimer la ressource externe quand la ressource managée est supprimée de Kubernetes.                                                                                             |

> | LateInitialize  | On parle de *late initialization* quand un controller Kubernetes met à jour les champs d'une resource après sa création. Cette police autorise Crossplane à mettre à jour la managed ressource avec des valeurs qui n'ont pas été renseignées pour sa création (e.g les ARNs)   |

Reviewer: ```suggestion
| LateInitialize  | On parle de *late initialization* quand [un controller Kubernetes](https://blog.eleven-labs.com/fr/un-premier-controleur-kubernetes-avec-controller-runtime/) met à jour les champs d'une resource après sa création. Cette police autorise Crossplane à mettre à jour la managed ressource avec des valeurs qui n'ont pas été renseignées pour sa création (e.g les ARNs)   |
```

|    ✔️     |         |                 |    ✔️     |    ✔️     | Crossplane ne supprime pas la ressource externe quand la ressource managée est supprimée. Crossplane n'importe aucun paramètres de la ressource externe.                           |
|    ✔️     |         |                 |    ✔️     |         | Crossplane créé la ressource externe mais n'applique aucun changement ni sur la ressource externe, ni sur la ressource managée et la suppression n'est pas autorisée                       |
|         |         |                 |    ✔️     |         | Crossplane observe seulement la resource.                                                                                                               |
|         |         |                 |           |         | Aucune managed policy, équivaut à mettre la ressource managée en "pause" (les modifications n'ont plus lieu).                                                                                                                        |

<br>

<div class="admonition tip" markdown="1"><p class="admonition-title">Important</p>

> Quand la fonctionnalité de managementPolicy est activée et qu'on utilise une police différente de celle par défault, la managementPolicy a priorité sur la deletionPolicy.

Reviewer: ```suggestion
Quand la fonctionnalité de managementPolicy est activée et qu'on utilise une police différente de celle par défaut, la managementPolicy a priorité sur la deletionPolicy.
```

</div>

Nous savons désormais créer des ressources unitaires et paramétrer des droits. Il est alors temps de créer une premiere abstraction, à l'image peut-être de ce que permettent les modules Terraform.

> ### Créer une première abstraction

Reviewer: ```suggestion
### Créer une première abstraction avec Crossplane
```

Attention, là, ça commence à se gâter. Certains termes sont assez proches et on a vite fait de les confondre. 
Pour mettre en place une abstraction, nous allons devoir combiner différents concepts sous-jacents dont l'interfaçage est représenté ci-dessous. Détaillons chaque éléments, en commençant par les compositions

![Logo Crossplane]({BASE_URL}/imgs/articles/2024-04-30-k8s-crossplane-introduction/schema_crossplane_end_users.svg)
<p style="text-align: center;">Schéma d'interfaçage entre les différents concepts Crossplane</p></br>

#### Compositions

> Une composition est sorte de liste de ressource que l'on pourra créé par l'intermédiaire d'un seul objet (que l'on appellera alors _ressource composite_, nous y reviendrons). Prenons l'exemple ci-dessous :

Reviewer: ```suggestion
Une composition est une sorte de liste de ressource que l'on pourra créer par l'intermédiaire d'un seul objet (que l'on appellera alors _ressource composite_, nous y reviendrons). Prenons l'exemple ci-dessous :
```

```yaml
apiVersion: apiextensions.crossplane.io/v1
kind: Composition
metadata:
  name: my-first-composition
spec:
  compositeTypeRef:
    apiVersion: aws.platformref.upbound.io/v1alpha1
    kind: XRDS
  resources:
    - name: RDSInstance
      base:
        apiVersion: rds.aws.upbound.io/v1beta1
        kind: Instance
        spec:
          forProvider:
            region: eu-west-3
    - name: SubnetGroup
      base:
        apiVersion: rds.aws.upbound.io/v1beta1
        kind: SubnetGroup     
        spec:
          forProvider:
            region: eu-west-3
            subnetIdRefs:
              - name: db-subnet1
              - name: db-subnet2
            tags:
              Name: My DB subnet group
    # etc.
```
On retrouve notre liste de ressources dans `spec.resources[]` dans lequel j'ai remis notre instance RDS déclarée au début de ce billet, accompagnée d'un subnet group. Notez que `spec.resources[].name` est purement arbitraire ici et permet uniquement de se repérer dans le template.

> Le bloc `spec.compositeTypeRef` est optionnel et permet d'exposer un endpoint d'API pour pouvoir réutiliser par référencer cette composition dans une autre composition. Dans l'exemple ci-dessous, un réutilise notre première composition en la référencant, et on déclare également un rôle IAM :

Reviewer: ```suggestion
Le bloc `spec.compositeTypeRef` est optionnel et permet d'exposer un endpoint d'API pour pouvoir réutiliser par référence cette composition dans une autre composition. Dans l'exemple ci-dessous, on réutilise notre première composition en la référençant, et on déclare également un rôle IAM :
```

            assumeRolePolicy: |
              {
                "Version": "2012-10-17",
                "Statement": [
                  {
                    "Effect": "Allow",
                    "Principal": {
                      "Service": "eks.amazonaws.com"
                    },
                    "Action": "sts:AssumeRole"
                  }
                ]
              }

    # etc.
```

Maintenant que nous avons une première composition, nous allons devoir mettre en place un endpoint d'API que nous pourrons ensuite utiliser pour créer nos ressources composites.

#### CompositeResourceDefinition

Pour cela, nous allons créer un objet `CompositeResourceDefinition` similaire à celui-ci :

```yaml
apiVersion: apiextensions.crossplane.io/v1
kind: CompositeResourceDefinition
metadata: 
  name: xrds.custom.api.exemple.org
spec:
  group: custom.api.exemple.org
  names:
    kind: XRDS
    plural: xrds
  versions:
    - name: v1alpha1
      schema:
        openAPIV3Schema:
          type: object
          properties:
            spec:
              type: object
              properties:
                type:
                  type: string
                  #etc.
```

> N.B : Par convention, Crossplane recommande pour les `CompositeResourceDefinition` de préfixer `spec.names[]` d'un `x` pour rappeller que la définition des APIs est custom (dans le sens qu'elle ne sont pas créés par le controller Crossplane mais géré par nous même).

Reviewer: ```suggestion
N.B : Par convention, Crossplane recommande pour les `CompositeResourceDefinition` de préfixer `spec.names[]` d'un `x` pour rappeler que la définition des APIs est custom (dans le sens qu'elles ne sont pas créées par le controller Crossplane mais géré par nous-même).
```

> Cet objet permet essentiellement de définir schéma d'API (type OpenAPI v3). Cette interface va nous permettre de definir la structure de nos ressource composites (type d'instance, moteur de base de donnée, région, etc. pour notre exemple) et `spec.group` et `spec.names.kind` nous permettrons de déterminer respectivement l'endpoint de l'API et le kind qu'on utilisera alors.s

Reviewer: ```suggestion
Cet objet permet essentiellement de définir le schéma d'API (type OpenAPI v3). Cette interface va nous permettre de définir la structure de nos ressources composites (type d'instance, moteur de base de donnée, région, etc. pour notre exemple) et `spec.group` et `spec.names.kind` nous permettrons de déterminer respectivement l'endpoint de l'API et le kind qu'on utilisera alors.
```

Dans cet exemple, nous aurons l'endpoint `custom.api.exemple.org` sur lequel nous pourrons créer une ressource composite avec kind `XRDS`. 

```yaml
apiVersion: custom.api.exemple.org/v1alpha1
kind: XRDS
metadata:
  name: my-composite-database
  annotations: 
    crossplane.io/external-name: my-custom-name
spec:
  type: large
```

#### Claims

> Par définition, les ressources composites (et les ressources managées associées) ne sont pas namespacées. Lorsqu'elle le sont, on parle alors de Claim. On le défini exactement comme une ressource composite à la différence près qu'elle n'utilise pas le même kind, qu'on définira dans `spec.claimNames`. 

Reviewer: ```suggestion
Par définition, les ressources composites (et les ressources managées associées) ne sont pas namespacées. Lorsqu'elles le sont, on parle alors de Claim. On le défini exactement comme une ressource composite à la différence près qu'il n'utilise pas le même kind, qu'on définira dans `spec.claimNames`. 
```

s

```yaml
apiVersion: apiextensions.crossplane.io/v1
kind: CompositeResourceDefinition
metadata: 
  name: xrds.custom.api.exemple.org
spec:
  group: custom.api.exemple.org
  names:
    kind: XRDS
    plural: xrds
  claimNames:
    kind: RDS
    plural: rds
  versions:
    - name: v1alpha1
      schema:
        openAPIV3Schema:
          type: object
          properties:
            spec:
              type: object
              properties:
                type:
                  type: string
                  #etc.
```

> On crééra par exemple ce Claim en utilisant le kind `RDS` (et non pas `XRDS`) et nos ressources seront cloisonnées au namespace `my-namespace`.

Reviewer: ```suggestion
On créera par exemple ce Claim en utilisant le kind `RDS` (et non pas `XRDS`) et nos ressources seront cloisonnées au namespace `my-namespace`.
```

```yaml
apiVersion: custom.api.exemple.org/v1alpha1
kind: RDS
metadata:
  name: my-claimed-database
  namespace: my-namespace
spec:
  type: large
```

A noter que `spec.claimNames` est un champ optionnel alors que `spec.names` est obligatoire !

> ### Objets "Usage"

Reviewer: ```suggestion
### Fonctionnalité objets "Usage" dans Crossplane
```

<div class="admonition tip" markdown="1"><p class="admonition-title">Note</p>

> <p>C'est une fonctionnalité encore en alpha qui est désactivée par défault. Pour l'utiliser, assurez-vous que le composant core de Crossplane soit lancé avec le flag `--enable-usages`</p>

Reviewer: ```suggestion
<p>C'est une fonctionnalité encore en alpha qui est désactivée par défaut. Pour l'utiliser, assurez-vous que le composant core de Crossplane soit lancé avec le flag `--enable-usages`</p>
```

</div>

Les objets `Usage` offrent un autre moyen d'empêcher la suppression d'une ressource. On précisera la référence de la ressource managée que l'on souhaite protégée dans un bloc `spec.of`

```yaml
apiVersion: apiextensions.crossplane.io/v1alpha1
kind: Usage
metadata:
  name: protect-prod-rds
spec:
  of:
    apiVersion: rds.aws.upbound.io/v1beta1
    kind: Instance
    resourceSelector:
      matchLabels:
        environment: production
  reason: "Production RDS instance- must not be deleted"
```

> Si on le souhaite, on peut autoriser la suppression de cette ressource qu'après avoir supprimé une autre ressource que l'on définiera dans un bloc `spec.by`. On fait ainsi un sorte d'ordonnancement pour les suppression. Il est tout à fait possible d'inclure ce type d'objets dans une composition

Reviewer: ```suggestion
Si on le souhaite, on peut autoriser la suppression de cette ressource après avoir supprimé une autre ressource que l'on définiera dans un bloc `spec.by`. On fait ainsi une sorte d'ordonnancement pour les suppressions. Il est tout à fait possible d'inclure ce type d'objets dans une composition.
```

```yaml
apiVersion: apiextensions.crossplane.io/v1
kind: Composition
metadata:
  name: my-first-composition
spec:
  compositeTypeRef:
    apiVersion: aws.platformref.upbound.io/v1alpha1
    kind: XRDS
  resources:
    - name: RDSInstance
      base:
        apiVersion: rds.aws.upbound.io/v1beta1
        kind: Instance
        metadata:
          name: my-rds-instance
        spec:
        # ...
    - name: SubnetGroup
      base:
        apiVersion: rds.aws.upbound.io/v1beta1
        kind: SubnetGroup     
        metadata:
          name: my-subnet-group
        spec:
        # ...
    # On inclut un objet "Usage" à notre composition
    - name: PreventDeletion
      base:
        apiVersion: apiextensions.crossplane.io/v1alpha1
        kind: Usage
        spec:
          of:
            apiVersion: rds.aws.upbound.io/v1beta1
            kind: SubnetGroup
            resourceRef:
              name: my-subnet-group
          by:
            apiVersion: rds.aws.upbound.io/v1beta1
            kind: Instance
            resourceRef:
              name: my-rds-instance
    # etc.
```

> ### Patches

Reviewer: ```suggestion
### Les Patches pour transformer les valeurs des champs des Claims dans Crossplane
```

![patch]({BASE_URL}/imgs/articles/2024-04-30-k8s-crossplane-introduction/patches_resized.png)

> Le mécanisme de patch permet de récupérer et de transformer les valeurs des champs de nos différentes *Claims* / ressources composites pour les injecter dans nos compositions. On pourra manipuler les valeurs renseignées par nous-même (e.g. taille d'un disque), ou injecté par le controller après création d'une ressource externe (e.g. ARN de la ressource). Un patch s'applique sur une ressources de composition et nous verrons plus tard un moyen pour l'appliquer à l'intégralité de la composition (via des `EnvironmentConfigs`).

Reviewer: ```suggestion
Le mécanisme de patch permet de récupérer et de transformer les valeurs des champs de nos différentes *Claims* / ressources composites pour les injecter dans nos compositions. On pourra manipuler les valeurs renseignées par nous-même (e.g. taille d'un disque), ou injecté par le controller après création d'une ressource externe (e.g. ARN de la ressource). Un patch s'applique sur une ressource de composition et nous verrons plus tard un moyen pour l'appliquer à l'intégralité de la composition (via des `EnvironmentConfigs`).
```

spec:
  compositeTypeRef:
    apiVersion: aws.platformref.upbound.io/v1alpha1
    kind: XRDS
  resources:
    - name: RDSInstance
      base:
        apiVersion: rds.aws.upbound.io/v1beta1
        kind: Instance
        spec:
          forProvider:
            region: eu-west-3
      patches:
        # on récupère le type d'instance depuis le Claim sans le transformer
        - type: FromCompositeFieldPath
          fromFieldPath: spec.type
          toFieldPath: spec.type         
        # On défini un pattern pour le nom de l'instance RDS qui contient un prefix statique suivi du type de l'instance. Le type d'instance et récupéré depuis le champ spec.type de notre Claim
        - type: FromCompositeFieldPath
          fromFieldPath: spec.type
          toFieldPath: annotations[crossplane.io/external-name]
          transforms:
            - type: string
                string:
                  type:
                    type: Format
                    fmt: "static-rds-name-%s" # résultat attendu avec notre exemple : "static-rds-name-large"
    - name: SubnetGroup
    # etc.
```

<div class="admonition warning" markdown="1"><p class="admonition-title">Attention</p>

> Il est tentant d'utiliser les champs `spec.resources[].metadata.name` pour nommer nos ressources. Toutefois, le comportement dans une composition n'est pas le même que dans une déclaration de ressource managée unitaire où `metadata.name` est alors le nom de la ressource externe par défaut. Ici l'utilisation de `spec.resources[].metadata.name` est bien autorisée mais les valeurs sont ignorées. Dans les compositions, c'est bien l'annotation "crossplane.io/external-name" qui va determiner le nom final.

Reviewer 2: Le markdown ne fonctionne pas car tu as oublié un saut de ligne après ton paragraphe

Author: OK, j'ai corrigé. Merci

  resources:
    - name: RDSInstance
      base:
        apiVersion: rds.aws.upbound.io/v1beta1
        kind: Instance
        spec:
          forProvider:
            region: eu-west-3
      patches:
        # On référence un PatchSet
        - type: PatchSet
          patchSetName: my-patchset        
        # On défini un pattern pour le nom de l'instance RDS qui contient un prefix statique suivi du type de l'instance. Le type d'instance et récupéré depuis le champ spec.type de notre Claim
        - type: FromCompositeFieldPath
          fromFieldPath: spec.type
          toFieldPath: annotations[crossplane.io/external-name]
          transforms:
            - type: string
                string:
                  type:
                    type: Format
                    fmt: "static-rds-name-%s" # résultat attendu avec notre exemple : "static-rds-name-large"
    - name: SubnetGroup
    # etc.
```

Je vous laisse explorer tout ce qu'il sera possible de faire en lisant la documentation sur le sujet : https://docs.crossplane.io/latest/concepts/patch-and-transform/

### Environment Config

<div class="admonition tip" markdown="1"><p class="admonition-title">Note</p>

> <p>`EnvironmentConfig` est une fonctionnalité actuellement en alpha et donc désactivée par défaut. Pour l'utiliser, assurez-vous que le composant core de crossplane soit lancé avec le flag `--enable-environment-configs` </p>

Reviewer 2: Le markdown ne fonctionne pas car tu as oublié un saut de ligne après ton paragraphe

Author: Il faut laisser la première ligne vide en effet et aussi laisser le texte brut (ne pas le mettre dans un paragraphe html)

kind: Composition
metadata:
  name: my-composition
spec:
  compositeTypeRef:
    apiVersion: aws.platformref.upbound.io/v1alpha1
    kind: XRDS
  environment:
    environmentConfigs:
      - type: Reference
        ref:
          name: stagging
    # On récupère le nom de l'environment dans l'objet `EnvironmentConfig` et on l'inject dans le nom de notre ressource composite 
    patches:
      - type: ToCompositeFieldPath
        fromFieldPath: env.name
        toFieldPath: annotations[crossplane.io/external-name]
        transforms:
          - type: string
              string:
                type:
                  type: Format
                  fmt: "static-rds-name-%s" # résultat attendu avec notre exemple : "static-rds-name-stag"
  resources:
    - name: RDSInstance
      base:
        apiVersion: rds.aws.upbound.io/v1beta1
        kind: Instance
        spec:
          forProvider:
            region: eu-west-3
        # etc.
```

Pour les patches déclarés dans `spec.environment.patches`, `ToCompositeFieldPath` copiera la donnée depuis le magasin vers la ressource composite et 
 `FromCompositeFieldPath` copiera la donnée depuis la composite resource vers le magasin.

> ### Functions

Reviewer: ```suggestion
### Les Functions de Crossplane
```

            base:
              apiVersion: rds.aws.upbound.io/v1beta1
              kind: Instance
              spec:
                forProvider:
                  region: eu-west-3
            # On déclare les patches ici mais en entrée d'une fonction cette fois
            patches:
              - type: FromCompositeFieldPath
                fromFieldPath: spec.type
                toFieldPath: spec.type
              - type: FromCompositeFieldPath
                fromFieldPath: spec.type
                toFieldPath: annotations[crossplane.io/external-name]
                transforms:
                  - type: string
                      string:
                        type:
                          type: Format
                          fmt: "static-rds-name-%s"
          - name: SubnetGroup
          # etc.
```

Avec cet exemple, nous nous attendons a obtenir le même resultat qu'avec la méthode précédente. La différence majeure est que le rendu du template ne repose plus uniquement sur le module core, mais sur différents packages que nous pourrons utiliser localement pour faire un _rendering_ de nos compositions à travers la commande `render` de la CLI Crossplane.

> ### Upjet

Reviewer: ```suggestion
### Upjet pour faciliter la création de providers Crossplane
```

Crossplane a dévoloppé le projet `Upjet` pour faciliter la création de providers Crossplane. Il est capable de génèrer automatiquement le code d'un provider Crossplane, en lisant le code d'un provider Terraform. Il a servi d'accélérateur pour la création des principaux providers Crossplane tels que `upbound/provider-aws` ; `upbound/provider-azure` ou `upbound/provider-gcp` et pourrait vous être utile si vous souhaitez migrer un de vos provider personnalisé.

Le dépôt du projet se trouve ici : https://github.com/crossplane/upjet

> ### Conclusion

Reviewer: ```suggestion
### Conclusion : à votre tour de vous lancer avec Crossplane
```
